### ISSUE TITLE
Marlin restarts on first layer for large parts only

### ISSUE THREAD

**Issue by LeavOA** [MarlinFirmware/Marlin#7510]
- Large Delta printer, Azteeg X3Pro Board.
- Description: Printing a large file (~200mm) causes the printer tocrash and restart during the first long straight line.
- Expected behaviour: no crashing! :)
- Actual behaviour: All startup procedures work find and the printer starts extruding the brim fine, but when the gcode reaches the first super long straight line, marlin crashes.
- Smaller files print fine :)

I'm guessing it might be some sort of timeout?

Log:
```
Recv: echo:G1 X-108.970 Y-38.035 E13.79160
Recv: echo:G1 X109.059 Y-38.035 E23.65284
Send: N13 M27*36
Recv: \x18\x00start
Recv: echo:Marlin 1.1.0
```

Configs:
[Leav.zip](https://github.com/MarlinFirmware/Marlin/files/1226140/Leav.zip)

Possible duplicate of #7303

**Comment by LeavOA** [comment]
possibly this line in Configuration.h has to do with it?

`#define EXTRUDE_MAXLENGTH 200`

Edit: Changing that line to 250 didn't solve that issue but modifying the part to not have a continuous straight line did make the problem go away for now....

any idea of the root cause for this?

**Comment by fiveangle** [comment]
What if you lower print speed in half ?  Does it still crash ?

**Comment by LeavOA** [comment]
Original speed was 50mm/s
Lowering speed to 25 mm/s made the line stop shorter (probably around same elapsed time)
Increasing speed to 100mm/s eliminated the problem again.

So I think this is a timeout issue.

**Comment by LeavOA** [comment]
I ended up digging into watchdog.c changed WDTO_4S to WDTO_8S which eliminated the symptoms completely for me without changing the print speed.

Though I am not familiar with the code base I would like to suggest that maybe the watchdog needs to reset every segment, instead of every G-Code line?

I am curious to hear your thoughts on the matter.

Thanks for this project!

**Comment by fiveangle** [comment]
Not too familiar with Delta's so someone else may have more useful suggestions.  Curious, returning to the original config, does altering this value change the extruded line length before crashing, or does it remain the same ?

```
  // Make delta curves from many straight lines (linear interpolation).
  // This is a trade-off between visible corners (not enough segments)
  // and processor overload (too many expensive sqrt calls).
  #define DELTA_SEGMENTS_PER_SECOND 200
```

**Comment by thinkyhead** [comment]
My guess is that the max speed is set too high for the CPU to keep up with the number of interrupts-per-second that would be required. If you lower your max feed rate you will no longer get crashing. This issue pops up more often when using 32X microstepping, because this effectively doubles the CPU load for every move.

You can also try lowering `DELTA_SEGMENTS_PER_SECOND`. This will produce fewer moves in the planner, but results in slightly rougher moves.

**Comment by LeavOA** [comment]
> If you lower your max feed rate you will no longer get crashing

I actually tried this, and the **opposite** was true.

> Original speed was 50mm/s
> Lowering speed to 25 mm/s made the line stop shorter (probably around same elapsed time)
> Increasing speed to 100mm/s eliminated the problem again.

I really think this is a watchdog issue for deltas.

Do cartesian printers use "segments" like deltas? if not that would explain why this hasn't come up before.
not to many deltas out there where you can issue a 200mm movement (and probably most of them are faster than my own... which is a bit slow)

bottom line: do you think maybe the issue could be that the watchdog isn't fed after every segment, but only after every g-code line?

**Comment by thinkyhead** [comment]
The watchdog only needs to be reset every 4 seconds, but is reset much more frequently — in fact, every time that the temperatures are read. The only reason the watchdog will not get reset within that 4 second time span is if the stepper interrupt is over-saturated with too many steps. Lowering your feedrates, segments-per-second, etc. will prevent the CPU from getting overloaded to the point where the Temperature::ISR fails to fire.

> Do cartesian printers use "segments" like deltas? if not that would explain why this hasn't come up before.

Cartesians frequently see this issue, especially on the Z axis when using too high a homing feedrate with 32x microstepping.

You can try using other optimization options, such as those which reduce the load of drawing the display. This is especially helpful when using a graphical display.

**Comment by pcarew** [comment]
@LeavOA, I believe that  you're spot on.
I just tripped over the same problem with a board reset in the first layer. I'm updating the FW for a 101Hero Delta printer. This is on release 1.1.0 RCBugFix.

Sure enough, changing the timeout from 4 to 8 seconds completely resolved my issue.

**Comment by wabbitguy** [comment]
I've been tripping over this as well since 1.1.4...the only way I could get it to print was to lower the Delta Segments, usually to below 190. So today, I was running segments at 196 and printing a plate of 160 x 75. It never made it past the printers outline prime. Just for the heck of it, I timed how long it took to do that one straight line and yep, 4 seconds. Printer stopped, MEGA rebooted.

I went into the watchdog.cpp changed WDTO_4S to WDTO_8S in the two places, then to really see if I could cause an issue, I bumped the Delta Segments back up to 200. I use 16x micro stepping.

@LeavOA I also think hit the nail on the head with the timeout. Because after changing that I'm printing perfectly again. No speed changes and increased delta segments.

Thus in spite of what the WDT should be doing, or programmed to be doing, on a delta it's a different kettle of fish. I would have never thought of the WDT...

**Comment by freeGIS4u** [comment]
Had the same problem on a custom Delta running Marlin 1.1.5.
First layer with 25% speed was fine, but Marlin crashed during printing the second one. Happend on straight lines longer than 15cm. 
Setting WDTO_4S to WDTO_8S fixed the problem, thanks a lot.

**Comment by AnHardt** [comment]
That implicates your heaters where not maintained for more than 4seconds because the processor was overloaded. Reducing delta segmentation, or print speed makes more sense.
(Don't fight the symptoms - fight the reasons)

**Comment by wabbitguy** [comment]
Reducing delta segments per second is the only way to sort of "fix" it without changing the WDT.

Reducing the print speed instantly makes it stall on long straight lines in 1.1.4 and newer. You have to increase speed to keep ahead of the 4 second timeout.

IMHO dropping the delta segments is the same sort of bandaid fix for the timeout as modifying the WDT. While lowering delta segments reduces the processor load and speeds up print times, the side effect is it lowers print quality via the speed increase. Thus when you reduce the print speed to regain quality, the timeout throws up again.

The WDT is a bandaid but unless the 4 second timer is worked into the software differently than it is now, one either runs old versions of Marlin, mods the WDT or puts up with lesser quality prints.

**Comment by pcarew** [comment]
@thinkyhead, can you point me to a description for the Marlin Github labels and their uses please.


I'm not sure why this issue has been closed as it's clearly a problem for some deltas.
Paul.

**Comment by LeavOA** [comment]
I'm also don't  understand why this was closed. Is this a "won't fix"?

**Comment by AnHardt** [comment]
@wabbitguy 
> Reducing delta segments per second is the only way to sort of "fix" it without changing the WDT.

You are right. For slower speeds `DELTA_SEGMENTS_PER_SECOND` is only producing smaller segments for a longer periode. Where almost all of them are thrown away by 'MIN_STEPS_PER_SEGMENT' early. That means planner buffer is running dry. That means 
```cpp
  // If the buffer is full: good! That means we are well ahead of the robot.
  // Rest here until there is room in the buffer.
  while (block_buffer_tail == next_buffer_head) idle();
```
idle() is not called here. What means 'thermalManager.manage_heater();', is not called, what would reset the WDT. 
The for-loop in 'prepare_kinematic_move_to()' what calls 'planner.buffer_line()` is never left until the G-command has finished. So there is no chance to call 'thermalManager.manage_heater()' at any other place.

So any DELTA move lasting longer than 'watchdog time out' is dammed to fail if it's not producing enough large enough segments to fill the planner buffer.

Short time workaround: Define `DELTA_SEGMENTS_PER_SECOND` small enough.
Possible long term fix: Call thermal  'thermalManager.manage_heater()', or better idle() or both, in the for loop of 'prepare_kinematic_move_to()', like we already do, for the same reason, in `plan_arc()`
```cpp
  millis_t next_idle_ms = millis() + 200UL;
...
for (uint16_t i = 1; i < segments; i++) { // Iterate (segments-1) times

    thermalManager.manage_heater();
    if (ELAPSED(millis(), next_idle_ms)) {
      next_idle_ms = millis() + 200UL;
      idle();
    }
```

That would,for 2.0.x-bugfix result in (motion.cpp):
```diff
@@ -592,10 +592,17 @@ float soft_endstop_min[XYZ] = { X_MIN_BED, Y_MIN_BED, Z_MIN_POS },
     // If there's only 1 segment, loops will be skipped entirely.
     --segments;
 
     // Calculate and execute the segments
     for (uint16_t s = segments + 1; --s;) {
+      static millis_t next_idle_ms = millis() + 200UL;
+      thermalManager.manage_heater();  // This returns immediately if not really needed.
+      if (ELAPSED(millis(), next_idle_ms)) {
+        next_idle_ms = millis() + 200UL;
+        idle();
+      }
+
       LOOP_XYZE(i) raw[i] += segment_distance[i];
       #if ENABLED(DELTA)
         DELTA_RAW_IK(); // Delta can inline its kinematics
       #else
         inverse_kinematics(raw);
```
@thinkyhead 
Would you please be so kind to check if there is a similar problem for scara-machines not handled here?

@all_other_interested
Please test the patch.

Regardless of this patch that moves will stutter if  `DELTA_SEGMENTS_PER_SECOND` is too large.
(~5h)

**Comment by wabbitguy** [comment]
@AnHardt I changed the WDT back 4S, added the patch below as you suggested to Marlin_Main.cpp.

I have an STL I sliced that's 188mm x 140mm square frame for testing. Set the Delta_Segments to 200 (if you're going to cause issues, my thinking is don't hold back)... set the speed to 20mm/s.

It takes just over 8 seconds to do one outside long edge of the frame. That should have easily caused the WDT problem.

Apart from the fact that the last ice age moved faster than my delta hot end does for this test, it's printing just fine (it's got an little over an hour to go so I'm just leaving it printing). I have travel moves set to 250mm/s and there's no issue with that.

I see no signs of any stutter during printing or moves so I'm not exactly sure where or when these "stutters" might expose themselves. Maybe they don't show at this glacially slow speed.

BTW - THANK YOU for the excellent description of the problem, and the possible fix! Much appreciated.
 ```
// Calculate and execute the segments
  millis_t next_idle_ms = millis() + 200UL;// MOD for WDT time out
  for (uint16_t s = segments + 1; --s;) {
    //
    thermalManager.manage_heater();
    if (ELAPSED(millis(), next_idle_ms)) {
      next_idle_ms = millis() + 200UL;
      idle();
    }
    //
    LOOP_XYZE(i) logical[i] += segment_distance[i];
#if ENABLED(DELTA)
    DELTA_LOGICAL_IK(); // Delta can inline its kinematics
#else
    inverse_kinematics(logical);
#endif

    ADJUST_DELTA(logical); // Adjust Z if bed leveling is enabled
```

**Comment by AnHardt** [comment]
@wabbitguy 
Thanks for the test.
You are on 1.1.x-bugfix?

Maybe the stuttering is very monotone and fast.
All segments in a single move have the same speed and the same length. So either they are produced faster then used, they are connected and the planner buffer runs full - then this problem is no problem. Or the segments are faster used than produced, the planer buffer contains never more than one line, there is nothing what the next segment could be connected to, speed is dropping to zero in between of the segments (that's usually called stuttering ).
Likely you have seen/heard now was a "endless" repeated sequence of 'MIN_STEPS_PER_SEGMENT' steps followed by a break where to 'short to plan' segments were produced until they added up to  'MIN_STEPS_PER_SEGMENT' steps.

To be honest - it souldn't have lasted that long until i got the idea for this fix. I searched an searched and searched - and when i discovered the problem i remembered the G2/G3 problem i fixed at least a year ago, with exactly the same reason - "subdivision of long lasting moves". Anyone a payable idea for improving recognition? ;-)

**Comment by AnHardt** [comment]
3:30 am here now. PRs tomorrow.

**Comment by wabbitguy** [comment]
@AnHardt I'm on 1.1.6, not the bug fix...6:30pm here...:-)

Can't help with improving anything, I'm retired...:-)

**Comment by AnHardt** [comment]
Ok. That part of the code hasn't changed.

**Comment by wabbitguy** [comment]
About the only thing I do notice is that some prints (vase mode) work better from the SD card than they do from Octoprint. My guess is there's a comm issue in there someplace but the work around is simple, I just copy to the SD card.

**Comment by wabbitguy** [comment]
BTW the print just finished and it worked perfectly so I'd say the patch was worth its weight in gold.

**Comment by AnHardt** [comment]
Octoprint is using conservative 'ping pong' - sends the next g-code only when have received an 'ok'. When not in  'ping pong' mode RepetierHost can better fill the buffers - but mind the buffer size - reduce it a bit in RH. (Not tested with above RH 2.0)

**Comment by fiveangle** [comment]
"some prints (vase mode) work better from the SD card than they do from Octoprint"

Be sure you do not have timelapse enabled for vase mode prints or Octoprint becomes starved for resources as it tries to capture/write a snapshot to the file system every z change, resulting in stuttering of all other actions.

-=dave

**Comment by wabbitguy** [comment]
I don't have a camera attached to Octoprint at the moment but thanks for the heads up on that, I'll check the config to make sure it's turned off in the config.

During some of my tests today in printing from Octoprint, I noticed that vase mode would leave artifacts so for the heck of it I swapped out the MEGA 2560 with a CH340G for a genuine MEGA with a 16U2. It didn't cure the artifacts, instead it made more of a pattern out of them as the vase mode built whereas with the CH340G MEGA they appeared more random. Weird.

I'm running a much faster print (70mm/s) at the moment and on long straight lines I can hear and see a very slight "stutter", I'll wait til it's done to see what it comes out like. Might toy around with that patch a bit to see where the sweet spot might lie.

**Comment by fiveangle** [comment]
Curious, do you have a repo with your complete current running code ?

**Comment by wabbitguy** [comment]
@fiveangle I'm using the 1.1.6 I downloaded directly from the code section. So no bug fixes or anything of the like in it, just pure stock. And I found the cause of the hiccup...it's the "idle()" call from that function patch. I nop'd it out and I'm running a fast test right now.

When completes, I'm going to slow it back down to glacial speed and see it trips the WDT on those long straight runs. According to my clock, I have about half hour left on the current print and then I'll shoot, once again, for worst case.

BTW, last repo I had was a '68 Dodge Charger...there, I dated myself....:-) I can put the code on my server if you'd like, no trouble at all in doing that.

**Comment by wabbitguy** [comment]
Just finished the test with some 200mm x 4mm strips running at a blistering 15mm/s. zzzzz

Took just under 11 seconds to do the long sides. No stutters, no WDT kick out..the only line of code I used in the patch is:

`thermalManager.manage_heater();`

I shall continue to test over the next few days and watch for gremlins.
